s3/s3_operations.cpp

```diff
@@ -245,20 +245,51 @@ namespace irods_s3 {
         }
 
-
-        // if this is a replication and we're the destination, get the data size from the source dataObjInfo
+        // special treatment for replication
+        // 1) data_size is only available from the REPLICATE_SRC entry so use that.
+        // 2) number_of_threads is available in REPLICATE_DEST entry so use that.
         if (oprType == REPLICATE_DEST) {
+            bool found_data_size = false;
+            bool found_number_of_threads = false;
 
             for (int i = 0; i < NUM_L1_DESC; ++i) {
+                const auto& l1d = L1desc[i];
+                const auto* dobj_input = l1d.dataObjInp;
+                const auto* dobj_info = l1d.dataObjInfo;
+
+                if (l1d.inuseFlag && dobj_input && dobj_input->objPath == file_obj->logical_path()) {
+                    // get the data size from source dataObjInfo
+                    if (!found_data_size && dobj_info && dobj_input->oprType == REPLICATE_SRC) {
+                        data_size = dobj_info->dataSize;
+                        rodsLog(developer_messages_log_level,
+                                "%s:%d (%s) [[%lu]] repl to s3 destination.  setting data_size to %zd\n",
+                                __FILE__,
+                                __LINE__,
+                                __FUNCTION__,
+                                thread_id,
+                                data_size);
+
+                        found_data_size = true;
+                    }
 
-                if (L1desc[i].inuseFlag && L1desc[i].dataObjInp && L1desc[i].dataObjInfo
-                       && L1desc[i].dataObjInp->objPath == file_obj->logical_path()
-                       && L1desc[i].dataObjInp->oprType == REPLICATE_SRC ) {
+                    // get the number_of_threads from destination dataObjInp
+                    if (!found_number_of_threads && dobj_input->oprType == REPLICATE_DEST) {
+                        number_of_threads = dobj_input->numThreads;
+                        rodsLog(developer_messages_log_level,
+                                "%s:%d (%s) [[%lu]] repl to s3 destination.  setting number_of_threads to %d",
+                                __FILE__,
+                                __LINE__,
+                                __FUNCTION__,
+                                thread_id,
+                                number_of_threads);
+
+                        found_number_of_threads = true;
+                    }
 
-                    data_size = L1desc[i].dataObjInfo->dataSize;
-                    rodsLog(developer_messages_log_level, "%s:%d (%s) [[%lu]] repl to s3 destination.  setting data_size to %zd\n",
-                            __FILE__, __LINE__, __FUNCTION__, thread_id, data_size);
-                    break;
+                    // once we have both pieces of information break out of for loop
+                    if (found_data_size && found_number_of_threads) {
+                        break;
+                    }
```

Comment:
If these are required to be true on the same iteration, then we need to add a check that enforces that requirement.

It can be as simple as checking that they are always equal just before or after this if-block.

Reply:
I changed the method to return an irods::error() in this case.  Please take a look at it.

Reply:
Actually these are not required to be true on the same iteration.  In fact they will never be true on the same iteration.

I would expect them to be true at the end but checking for these both be be true at the end may have broken a test which means that something didn't behave the way I expected.  I am looking into it now.

Reply:
It appears that on an irepl -r there is not a replication source in the `L1desc` table.  Therefore, I can't guarantee that the `found_data_size` will be true.

This is acceptable because if the data size is not known it will default to using a cache file.  I kept in the check to make sure `found_number_of_threads` was true but took out the one for `found_data_size`.

I am running it through some tests and if it passes I will commit the latest change.

Reply:
Ok. Will wait to see what you find out.

Reply:
Running one suite of cacheless tests passed.  This should be good to go. 

Note that in the scenario I referenced, the destination L1desc entry does have a srcL1descIdx but the actual entry at that index is not valid or is no longer valid.